### PR TITLE
Resolve performance degradations from CP PR

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This resolves performance degradation introduced in #2570.

image

## Changes proposed in this PR:
- Reimplement Var/Param `__getitem__()` to reduce overhead for the most common (math programming) situations
- Reduce isinstance calls in the `BoundInitializer`
- Update implementation of `BoundInitializer` to  make it behave more like `Initializer()`
- Update docuumentation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/initializer.py [hunk 1]
```diff
@@ -445,43 +469,71 @@ def indices(self):
...
+    Note that this class is meant to mimic the behavior of
+    :py:func:`Initializer` and will return ``None`` if the initializer
+    that it is wrapping is ``None``.
+
+    Parameters
+    ----------
+    arg:
+
+        As with :py:func:`Initializer`, this is the raw argument passed
+        to the compnent constructor.
```
Comment (blnicho) [inline pyomo/core/base/initializer.py:487]: Typo: component

File: pyomo/core/base/initializer.py [hunk 2]
```diff
@@ -445,43 +469,71 @@ def indices(self):
...
+        indexed and the sequence passed to the initializer is not of
+        length 2
+
+    """
     __slots__ = ('_initializer',)
 
-    def __init__(self, obj, init):
-        if obj.is_indexed():
+    def __new__(cls, arg=None, obj=NOTSET):
+        # The Initialier() function returns None if the initializer is
```
Comment (blnicho) [inline pyomo/core/base/initializer.py:500]: Typo: Initializer

**Review by blnicho — APPROVED** [review]
I found a couple typos but otherwise this looks fine.

File: pyomo/core/base/initializer.py [hunk 3]
```diff
@@ -45,21 +43,45 @@ def Initializer(init,
     function standardizes the processing of keyword arguments and
     returns "initializer classes" that are specialized to the specific
     data type provided.
+
+    Parameters
+    ----------
+    arg:
+
+        The argument passed to the component conostructor.  This could
```
Comment (emma58) [inline pyomo/core/base/initializer.py:51]: Typo: constructor

**Review by emma58 — APPROVED** [review]

**Pull request merged**